### PR TITLE
Add support to validate the legal name against the payout method

### PR DESCRIPTION
**Pull request by SudharakaP**
Related to https://github.com/opencollective/opencollective/issues/4581

Related Front-end PR: https://github.com/opencollective/opencollective-frontend/pull/6818

image

### CODE REVIEW

File: server/models/PayoutMethod.ts [hunk 1]
```diff
@@ -23,11 +23,13 @@ export enum PayoutMethodTypes {
 /** An interface for the values stored in `data` field for PayPal payout methods */
 export interface PaypalPayoutMethodData {
   email: string;
+  accountHolderName: string;
```
Comment (Betree) [inline server/models/PayoutMethod.ts:26]: Why adding this to `PaypalPayoutMethodData` and `OtherPayoutMethodData`? Don't we want that just for `BankAccountPayoutMethodData`?

Comment (SudharakaP) [reply server/models/PayoutMethod.ts:26]: @Betree : The reason this was added is due to typescript errors such as; 

```
server/graphql/common/expenses.ts:632:26 - error TS2339: Property 'accountHolderName' does not exist on type 'PayoutMethodDataType'.
  Property 'accountHolderName' does not exist on type 'PaypalPayoutMethodData'.
```

I couldn't find a more elegant way to do this. 🤔

Comment (Betree) [reply server/models/PayoutMethod.ts:26]: The property doesn't exists on `PaypalPayoutMethodData` and `OtherPayoutMethodData`, so there's no reason to add it there. What we want is to make sure typescript understands that we're playing with a `BankAccountPayoutMethodData` in there. For example, by casting the value:

```ts
if (payoutMethod?.type === PayoutMethodTypes.BANK_ACCOUNT) {
  const payoutMethodData = <BankAccountPayoutMethodData>payoutMethod.data;
  if (payoutMethodData?.accountHolderName !== expenseData.fromCollective.legalName) {
    ...
  }
}
```

Comment (SudharakaP) [reply server/models/PayoutMethod.ts]: Ah, that makes perfect sense. I need to brush up on my Typescript skills. 🤔

**Review by Betree — APPROVED** [review]
Looks good to me

**Review by Betree — COMMENTED** [review]
Actually, can we make the check case-insensitive? Would be frustrating for users to be blocked because they've made their family name uppercase in one of the forms

**Comment by SudharakaP** [comment]
> Actually, can we make the check case-insensitive? Would be frustrating for users to be blocked because they've made their family name uppercase in one of the forms

Thanks for pointing it out. I just made some checks to make sure we perform the check case-insensitively. 😄

**Review by Betree — COMMENTED** [review]
I wanted to estimate how this change is going to impact users. I did that using this query:
```sql
SELECT MAX(e."createdAt") AS "createdAt", payee.name, pm."data" ->> 'accountHolderName'
FROM "Expenses" e
INNER JOIN "PayoutMethods" pm ON e."PayoutMethodId" = pm.id
INNER JOIN "Collectives" payee ON e."FromCollectiveId" = payee.id
WHERE pm."type" = 'BANK_ACCOUNT'
AND payee.name NOT ILIKE pm."data" ->> 'accountHolderName'
GROUP BY payee.name, pm."data" ->> 'accountHolderName'
ORDER BY MAX(e."createdAt") DESC
```

Not sharing the real names for privacy reasons, but here are a few cases that I think we should handle from the start:
- Trim `pm."data" ->> 'accountHolderName'`, as it's apparently not. Ideally we should do that on the model and have a migration, but if you want to go fast adding a `.trim()` before the comparison is also fine.
- Special characters: `François` VS `Francois`. Bank accounts sometimes don't support special characters like that, so we should ideally run both strings through `str.normalize("NFD").replace(/[\u0300-\u036f]/g, "")`
- `Evil Corp, Inc` -> `Evil Corp, Inc.` (the dot, doesn't sound like this should fail)
- First name/last name order: `Benjamin Piouffle` -> `Piouffle Benjamin`. Would be great to ignore the order of the words in the comparison.

Because of the added complexity, I would actually move all these checks to a separate function that loosely compares two names and returns a boolean to say whether they're egual.

Also, we're missing the validation when editing an expense. It should apply to both create and edit.

**Comment by SudharakaP** [comment]
> I wanted to estimate how this change is going to impact users. I did that using this query:
> 
> ```sql
> SELECT MAX(e."createdAt") AS "createdAt", payee.name, pm."data" ->> 'accountHolderName'
> FROM "Expenses" e
> INNER JOIN "PayoutMethods" pm ON e."PayoutMethodId" = pm.id
> INNER JOIN "Collectives" payee ON e."FromCollectiveId" = payee.id
> WHERE pm."type" = 'BANK_ACCOUNT'
> AND payee.name NOT ILIKE pm."data" ->> 'accountHolderName'
> GROUP BY payee.name, pm."data" ->> 'accountHolderName'
> ORDER BY MAX(e."createdAt") DESC
> ```
> 
> Not sharing the real names for privacy reasons, but here are a few cases that I think we should handle from the start:
> 
>     * Trim `pm."data" ->> 'accountHolderName'`, as it's apparently not. Ideally we should do that on the model and have a migration, but if you want to go fast adding a `.trim()` before the comparison is also fine.
> 
>     * Special characters: `François` VS `Francois`. Bank accounts sometimes don't support special characters like that, so we should ideally run both strings through `str.normalize("NFD").replace(/[\u0300-\u036f]/g, "")`
> 
>     * `Evil Corp, Inc` -> `Evil Corp, Inc.` (the dot, doesn't sound like this should fail)
> 
>     * `Evil Corp Inc` -> `Evil Corp, Inc` (the comma, doesn't sound like this should fail). Maybe we should ignore punctuation altogether
> 
>     * First name/last name order: `Benjamin Piouffle` -> `Piouffle Benjamin`. Would be great to ignore the order of the words in the comparison.
> 
> 
> Because of the added complexity, I would actually move all these checks to a separate function that loosely compares two names and returns a boolean to say whether they're equal. We should also unit test it, using all the examples provided above.
> 
> Also, we're missing the validation when editing an expense. It should apply to both create and edit.

@Betree : Thanks for the detailed explanation. I've added support for all the points mentioned here using the [`localeCompare` ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare) function. All punctuation is ignored and if the name consist only of two parts (first name and last name) we now make sure the order is ignored.

https://github.com/opencollective/opencollective-api/blob/55302337bf3a1fee17ea400b643dcfcdeeb2ec58/server/graphql/common/expenses.ts#L715-L744

Let me know how it looks. 😉

File: test/server/graphql/common/expenses.test.js [hunk 2]
```diff
@@ -407,5 +408,13 @@ describe('server/graphql/common/expenses', () => {
       expect(await canUnschedulePayment(collectiveAccountantReq, expense)).to.be.false;
       expect(await canUnschedulePayment(hostAccountantReq, expense)).to.be.false;
     });
+
+    it('make sure legal name is validated against the account holder name', async () => {
+      expect(isAccountHolderNameAndLegalNameMatch('Evil Corp, Inc', 'Evil Corp, Inc.')).to.be.true;
+      expect(isAccountHolderNameAndLegalNameMatch('François', 'Francois')).to.be.true;
+      expect(isAccountHolderNameAndLegalNameMatch('Sudharaka Palamakumbura', 'Palamakumbura Sudharaka')).to.be.true;
```
Comment (Betree) [inline test/server/graphql/common/expenses.test.js:415]: Let's add some more tests to increase the resilience here
```suggestion
      expect(isAccountHolderNameAndLegalNameMatch('Evil Corp, Inc', 'Evil Corp, Inc.')).to.be.true;
      expect(isAccountHolderNameAndLegalNameMatch('Evil Corp Inc', 'Evil Corp, Inc.')).to.be.true;
      expect(isAccountHolderNameAndLegalNameMatch(' Evil   Corp,    Inc.', '   Evil Corp   Inc')).to.be.true;
      expect(isAccountHolderNameAndLegalNameMatch('François Dêaccènt', 'Francois DeAccEnt')).to.be.true;
      expect(isAccountHolderNameAndLegalNameMatch('Sudharaka Palamakumbura', 'Palamakumbura Sudharaka')).to.be.true;
      expect(isAccountHolderNameAndLegalNameMatch('Sudharaka Palamakumbura', 'Sudharaka Palamakumbura')).to.be.true;
```

File: server/graphql/common/expenses.ts [hunk 3]
```diff
@@ -709,6 +719,39 @@ export const getItemsChanges = async (
...
+        sensitivity: 'base',
+        ignorePunctuation: true,
+      }) &&
+      accountHolderName.localeCompare(legalNameReversed, undefined, {
+        sensitivity: 'base',
+        ignorePunctuation: true,
+      })
+    ) {
+      return false;
+    }
```
Comment (Betree) [inline server/graphql/common/expenses.ts:750]: :ice_cream: A small helper could improve readability here

```suggestion
    const localeCompare = (str1, str2) => str1.localeCompare(str2, undefined, { sensitivity: 'base', ignorePunctuation: true });
    if (localeCompare(accountHolderName, legalName) || localeCompare(accountHolderName, legalNameReversed)) {
      return false;
    }
```

File: server/graphql/common/expenses.ts [hunk 4]
```diff
@@ -786,6 +829,19 @@ export async function editExpense(
...
+
+  // Validate bank account payout method
+  if (payoutMethod?.type === PayoutMethodTypes.BANK_ACCOUNT) {
+    const payoutMethodData = <BankAccountPayoutMethodData>payoutMethod.data;
+    if (
+      !isAccountHolderNameAndLegalNameMatch(
+        payoutMethodData?.accountHolderName,
+        <string>expenseData.fromCollective.legalName,
+      )
+    ) {
```
Comment (Betree) [inline server/graphql/common/expenses.ts:841]: :ice_cream: I would personally keep the helper function just about comparing name (also to make it consistent with the frontend), and add a check here:


```suggestion
    const accountHolderName = payoutMethodData?.accountHolderName;
    const legalName = expenseData.fromCollective.legalName;
    if (accountHolderName && legalName && !namesAreEqual(accountHolderName, legalName)) {
```

**Review by Betree — COMMENTED** [review]
Great tests here! I suggested some more to make it more resilient, the rest are just suggestions on code style that are personal preferences; feel free to ignore them.

File: server/graphql/common/expenses.ts (hunk 3 above)
Comment (SudharakaP) [reply server/graphql/common/expenses.ts]: Done. 👍🏽

File: test/server/graphql/common/expenses.test.js (hunk 2 above)
Comment (SudharakaP) [reply test/server/graphql/common/expenses.test.js:415]: Done. 👍🏽

**Comment by SudharakaP** [comment]
> Great tests here! I suggested some more to make it more resilient, the rest are just suggestions on code style that are personal preferences; feel free to ignore them.

Thanks for the review. I've addressed all of it. Let me know if anything further needs to be done. 👍🏽

**Review by Betree — APPROVED** [review]

**Pull request merged**